generic-ec-zkp/src/dlog_eq.rs

```diff
@@ -0,0 +1,277 @@
+//! # fn send<T>(_: T) {}
+//!
+//! // `x` is a secret discrete logarithm only known to prover
+//! let x = SecretScalar::<E>::random(&mut OsRng);
+//!
+//! // `X` is a public point corresponding to `x`
+//! let X = Point::generator() * &x;
+//! // `H` is some point known to both parties
+//! let H = Point::generator() * Scalar::random(&mut OsRng);
+//! // `Z` is the object of the proof
+//! let Z = H * &x;
+//!
+//! let data = non_interactive::Data {
+//!     pub_share: X,
+//!     base: H,
+//!     exp: Z,
+//! };
+//!
+//! // Prover chooses a private nonce
+//! let r = Scalar::random(&mut OsRng);
+//! // Prover proves in zero knowledge the equality of `H * x = Z`
+//! let proof = non_interactive::prove::<sha2::Sha256, E>(&"shared_state", &x, data, r);
+//! // The proof is sent to the Verifier
+//! send(proof);
+//!
+//! // Verifier checks that the proof is correct
+//! non_interactive::verify::<sha2::Sha256, E>(&"shared_state", data, proof)
+//!     .expect("Verification failed!");
+//! ```
+//!
+//! ## Interactive example
+//!
+//! ```rust
+//! # use generic_ec::{Curve, Scalar, SecretScalar, Point};
+//! # use generic_ec_zkp::dlog_eq::interactive;
+//! # use rand::rngs::OsRng;
+//! # use generic_ec::curves::Secp256k1 as E;
+//! # fn send<T>(_: T) {}
+//!
+//! // `x` is a secret discrete logarithm only known to prover
+//! let x = SecretScalar::<E>::random(&mut OsRng);
+//!
+//! // `X` is a public point corresponding to `x`
+//! let X = Point::generator() * &x;
+//! // `H` is some point known to both parties
+//! let H = Point::generator() * Scalar::random(&mut OsRng);
+//! // `Z` is the object of the proof
+//! let Z = H * &x;
+//!
+//! let data = interactive::Data {
+//!     pub_share: X,
+//!     base: H,
+//!     exp: Z,
+//! };
+//!
+//! // Prover chooses a nonce for itself
+//! let r = Scalar::random(&mut OsRng);
+//! // Prover commits to the data
+//! let commitment = interactive::commit_data(data, r);
+//! // Prover sends this commitment to the verifier
+//! send(commitment);
+//!
+//! // Verifier sends a challenge to the prover
+//! let challenge = Scalar::random(&mut OsRng);
+//! send(challenge);
+//!
+//! // Prover receives the challenge and computes the proof of equality `H * x = Z`
+//! let proof = interactive::prove(r, challenge, &x);
+//! // This proof is sent to the verifier
+//! send(proof);
+//!
+//! // Verifier checks that the proof is correct
+//! interactive::verify(data, commitment, challenge, proof)
+//!     .expect("Verification failed!");
+//! ```
+
+use generic_ec::{Curve, Point};
+
+/// Object of the proof
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "udigest", derive(udigest::Digestable), udigest(bound = ""))]
+pub struct Data<E: Curve> {
+    /// `X`, the value inside `log_G`
+    pub pub_share: Point<E>,
+    /// `H`, the base for the second logarithm
+    pub base: Point<E>,
+    /// `Z`, the value inside the second logarithm `log_H`
+    pub exp: Point<E>,
+}
```

Comment:
So here, for sake of generality, it's worth adding the second base, and change naming perhaps (why `exp`, we use multiplicative notation!)

```suggestion
pub struct Data<E: Curve> {
    /// Generator $H_1$
    pub base1: Point<E>,
    /// Generator $H_2$
    pub base2: Point<E>,
    /// Point $P_1 = x_1 \cdot H_1$ such that $x_1$ is only known to the prover
    pub point1: Point<E>,
    /// Point $P_2 = x_2 \cdot H_2$ such that $x_2$ is only known to the prover
    pub point2: Point<E>,
}
```

Reply:
You're right, I missed that part. Though I wonder if this generality is worth worse ergonomics in the common case where g is just the habitual generator. I just had an idea how to make a smart constructor to return the ergonomics back..

Reply:
Yeah the constructor may work

Reply:
Did the constructor in an earlier commit, have a look